components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt

```diff
@@ -260,7 +258,11 @@ class DownloadsFeature(
         appChooserDialog.setApps(apps)
         appChooserDialog.onAppSelected = { app ->
             if (app.packageName == applicationContext.packageName) {
-                startDownload(download)
+                if (applicationContext.isPermissionGranted(downloadManager.permissions.asIterable())) {
+                    startDownload(download)
+                } else {
+                    onNeedToRequestPermissions(downloadManager.permissions)
```

Comment:
There is a bug  🐛  here,  if you see the last line of this function [useCases.consumeDownload(tab.id, download.id)](https://github.com/mozilla-mobile/android-components/blob/9b35fb98ab182b5d8f7985d35648797a67197fb4/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt#L277) is consuming the download request either if the download is triggered (`startDownload`)  or if we are showing the permissions dialog (`onNeedToRequestPermissions`), causing the first time that we try to download with an external app and select this browser the download doesn't start, after the permissions dialog is shown and the permission granted. This is because we consumed the download request on  useCases.consumeDownload(tab.id, download.id)](https://github.com/mozilla-mobile/android-components/blob/9b35fb98ab182b5d8f7985d35648797a67197fb4/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt#L277).

This structure is because before we used to warranty that at this point we had permissions granted. Be aware that when fixed the issue above you will see another one,  after you grant the permission the dialog is going to show again instead of starting the download, this is because after the permission is granted we call [processDownload](https://github.com/mozilla-mobile/android-components/blob/9b35fb98ab182b5d8f7985d35648797a67197fb4/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt#L205).

Reply:
Hmm, then we should add consumeDownload to both below L262 and L269 and remove from L277.

Also for processDownload we can change it with startDownload because permission already granted and user selected fenix. We can directly start download.

@Amejia481 Is this okey?

Reply:
> Hmm, then we should add consumeDownload to both below L262 and L269 and remove from L277.

That sounds fair, only one change instead of consuming on L269 do it on L275 just after the `catch` statement, we need to update the test to verify that we are consuming in both cases.


> Also for processDownload we can change it with startDownload because permission already granted and user selected fenix. We can directly start download. Because we'll just want permission if user selects fenix in list and soo we can directly start download in onPermissionResult.

That's ok, we just need a way to remember that the user already selected the app to be the downloader.

Reply:
> we need to update the test to verify that we are consuming in both cases.

`None exception is thrown when unable to open an app third party for downloading` and `when an app third party is selected for downloading we MUST forward the download` and `when our app is selected for downloading and permission granted then we should perform the download` checks consumeDownload so don't need to update these tests, only fixed the one i added.